### DISCUSSION TITLE
mcp251xfd strange interrupt configuration

### DISCUSSION THREAD

**Discussion by adri1mart1** [zephyrproject-rtos/zephyr#74561] [Q&A] [answered]
Hello zephyr community,

I'm playing with the [mcp251863-click](https://www.mikroe.com/mcp251863-click) board and with an EFR32MG24.
I'm using the [can_mcp251xfd](https://github.com/zephyrproject-rtos/zephyr/blob/97a97c744a294ba14278a1f436d46eaa2d488354/drivers/can/can_mcp251xfd.c) driver and one thing really bother me out.

In the init function, the interrupt is configured using `GPIO_INT_LEVEL_ACTIVE`:

```
if (gpio_pin_interrupt_configure_dt(&dev_cfg->int_gpio_dt, GPIO_INT_LEVEL_ACTIVE) < 0) {
	return -EINVAL;
}
```

However, on the EFR32, the gpios cannot be configured using level interrupts, [see gpio driver](https://github.com/zephyrproject-rtos/zephyr/blob/14642091730b26abf49e9ca359fe993ac8f56137/drivers/gpio/gpio_gecko.c#L288):

```
/* Interrupt on static level is not supported by the hardware */
if (mode == GPIO_INT_MODE_LEVEL) {
	return -ENOTSUP;
}
```

By curiosity, I've checked on stm32 and it seems it does not support it as well: --> [see driver](https://github.com/zephyrproject-rtos/zephyr/blob/14642091730b26abf49e9ca359fe993ac8f56137/drivers/gpio/gpio_stm32.c#L651)

My question is:

- Am I the first one ever who have tried to connect an STM32 (or EFR32 in my case) with an MCP251x CAN controller ?
- Should I update the existing driver to be compatible with EDGE interrupt instead of LEVEL interrupts ?

Thanks,
Adrien M.

**Comment by ndrs-pst** [comment]
Ping @talih0, CC @henrikbrixandersen.
I'm curious too 😄.

**Comment by henrikbrixandersen** [comment]
> * Should I update the existing driver to be compatible with EDGE interrupt instead of LEVEL interrupts ?

How would you change it? The MCP2518FD uses level-triggered interrupt lines (meaning the interrupt line is asserted as long as an interrupt flags is present). This is different from edge-triggered interrupts.

**Reply by adri1mart1** [reply]
I am definitely not sure because I'm not totally familiar with edge/level interrupt, however, I would probably think about switching the interrupt mode to EDGE, and force the driver to reprocess the code it needs to process if the interrupt is still active after a first process.

How would you do to setup a STM32 / EFR32 with this chip otherwise ?

**Reply by henrikbrixandersen** [reply]
>  ... and force the driver to reprocess the code it needs to process if the interrupt is still active after a first process.

So basically treating a level-triggered interrupt as an edge-triggered interrupt. That's not a good solution, as it can lead to races in the interrupt handling code.

**Reply by adri1mart1** [reply]
I agree, however this is the only solution I can see right now. Any though on another one ?

**Reply by henrikbrixandersen** [reply]
Yes. Use hardware interfaces that match each other. If you need level-triggered GPIO interrupt support for a peripheral, use a microcontroller with support for level-triggered GPIO interrupts.

**Comment by ndrs-pst** [comment]
@henrikbrixandersen For example, STM32 also only supports edge-triggered interrupts (see [this link](https://community.st.com/t5/stm32-mcus-products/how-to-implement-level-triggered-interrupts-with-exti/td-p/65224)).
So if the driver itself requires `GPIO_INT_LEVEL_ACTIVE`, it will hinder the usage with MCUs that only support edge-triggered interrupts.

Any chance that the driver provides the option to enable usage with MCUs that only support edge-triggered interrupts?
I may ping @erwango for some opinion about this.

**Reply by henrikbrixandersen** [reply]
You seem to focus on the GPIO controller _driver_ not supporting level-triggered interrupts, but if the GPIO controller  _hardware_ does not support level-triggered interrupts, how do you expect to be able to use a level-triggered interrupt with that GPIO controller? We cannot change hardware capabilities in software, no matter how hard we try.

**Reply by ndrs-pst** [reply]
I understand your point. For STM32, when configured to both-edge EXTI, the hardware doesn't provide specific rising or falling edge information, so we need to read the GPIO to determine the current state, then find out if it's rising or falling.

Maybe the user can decide to use `GPIO_INT_EDGE_FALLING` via KConfig while accepting the risk of missing interrupts. ?

**Reply by henrikbrixandersen** [reply]
> I understand your point. For STM32, when configured to both-edge EXTI, the hardware doesn't provide specific rising or falling edge information, so we need to read the GPIO to determine the current state, then find out if it's rising or falling.

That's racy too. 

> Maybe the user can decide to use `GPIO_INT_EDGE_FALLING` via KConfig while accepting the risk of missing interrupts. ?

Definitely not.

**Comment by adri1mart1** [comment, marked as answer]
Hello all,

An update on this topic. In the end, I finally choose a solution in which I completely remove the interrupt and use
polling of the pin. I agree with you henrik, this is not a good way to do it. However, for now, this is a solution
that suits my needs. 

I could not easily change the hardware, I don't need important CAN bitrate so that's fine. I don't think there's 
any change to bring to zephyr code, it's quite easy to spot when you're using the hardware that you have a misconfiguration
using GPIO_LEVEL_INTERRUPTS. 

Regarding one of my initial question: "Am I the first one ever who have tried to connect an STM32 (or EFR32 in my case) with an MCP251x CAN controller ?"

The answer is (probabbly) yes with this particular configuration. According to some discussion with the HW team, this 
MCP251862 component is mostly used with MPUs, when dealing with MCUs, most of the time you choose an MCU that has
internal CAN controler, or you may choose the MCP2515 which is not concerned by that. 

I'm closing this topic, thanks for you explanations
